Add data stream docs to Fleet user guide (#429)
elastic/observability-docs: docs/en/ingest-management/data-streams.asciidoc

[[data-streams]]
= Data streams

> {agent} uses data streams to easily store append-only time series data across multiple indices
> while giving you a single named resource for requests.

Reviewer: ```suggestion
{agent} uses data streams to store time series data across multiple indices
while giving you a single named resource for requests.
```

Reviewer: My brain tripped over this sentence a couple of times. It's maybe too packed.  Is the term append-only needed?

Author: You're right; we don't need that level of detail here.

Data streams are well-suited for logs, metrics, traces, and other continuously generated data.
They offer a host of benefits over other indexing strategies:

* *Reduced number of fields per index*: Indices only need to store a specific subset of your
data–meaning no more indices with hundreds of thousands of fields.
This leads to better space efficiency and faster queries.
As an added bonus, only relevant fields are shown in Discover.

* *More granular data control*: For example, filesystem, load, cpu, network, and process metrics are sent

> to different indices–each potentially with its own rollover, retention, and security permissions.

Reviewer: ```suggestion
to different indices–-each potentially with its own rollover, retention, and security permissions.
```

Reviewer: I think -- gets converted into an em dash.

Author: That's already an em dash :)

Author: It's hard to tell in the source file, but a little easier in the built version:
<img width="685" alt="Screen Shot 2021-03-10 at 7 17 13 PM" src="https://user-images.githubusercontent.com/5618806/110730393-3d378d00-81d5-11eb-9885-9de1a9841939.png">

* *Flexible*: Use the custom namespace component to divide and organize data in a way that

> makes sense to your use-case or company.

Reviewer: ```suggestion
makes sense to your use case or company.
```

* *Fewer ingest permissions required*: Data ingestion only requires permissions to append data.

[discrete]
[[data-streams-naming-scheme]]
== Data stream naming scheme

{agent} uses the Elastic data stream naming scheme to name data streams.
The naming scheme splits data into different streams based on the following components:

`type`::
A generic `type` describing the data, such as `logs`, `metrics`, `traces`, or `synthetics`.
// Corresponds to the `data_stream.type` field.

`dataset`::
The `dataset` is defined by the integration and describes the ingested data and its structure for each index.
For example, you might have a dataset for process metrics with a field describing whether the process is running or not,
and another dataset for disk I/O metrics with a field describing the number of bytes read.
+
For APM data, the `dataset` is the instrumented service's `service.name`.

`namespace`::
A user-configurable arbitrary grouping, such as an environment (`dev`, `prod`, or `qa`),
a team, or a strategic business unit.

> A `namespace` can be up to 100 bytes in length (multi-byte characters will count toward this limit faster).

Reviewer: ```suggestion
A `namespace` can be up to 100 bytes in length (multibyte characters will count toward this limit faster).
```

Reviewer: Both forms are correct, but I think all one word is in the spirit of Elastic style.

Using a namespace makes it easier to search the data from a given source by using index patterns, or to give users permissions to data by assigning an index pattern to user roles.
// Corresponds to the `data_stream.dataset` field.

The naming scheme separates each components with a `-` character:

[source,text]
--
<type>-<dataset>-<namespace>
--

> For example, if you've set up the Nginx integration with a namespace of `prod`,

Reviewer: You might want to format this differently to make it easier to scan. Maybe more white space? Honestly I think the line wraps in the middle of names are making it harder to follow the example.

Author: I see what you mean.

Author: I added additional white space and moved the examples to code blocks in 2ddd984.

{agent} uses the `logs` type, `nginx.access` dataset, and `prod` namespace to store data in the
`logs-nginx.access-production` data stream.
If you use the APM integration with a namespace of `dev`, and `service.name` (dataset) of `frontend`,
{agent} uses the `traces-apm.frontend.dev` data stream.

All data streams, and the pre-built dashboards that they ship with,
are viewable on the Fleet Data Streams page:

[role="screenshot"]
image::images/kibana-fleet-datastreams.png[Data streams page]

TIP: If you're familiar with the concept of indices, you can think of each data stream as a separate index in {es}.
Under the hood though, things are a bit more complex.

> All of the juicy details are available in {ref}/data-streams.html[{es} Data streams].

Reviewer: Nice! Knowing "juicy details" is way more intriguing than having to read some other docs. :->
